src/librustc_mir/shim.rs

```diff
@@ -99,7 +99,13 @@ fn make_shim<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         ty::InstanceDef::DropGlue(def_id, ty) => {
             build_drop_shim(tcx, def_id, ty)
         }
-        ty::InstanceDef::CloneCopyShim(def_id, ty) |
+        ty::InstanceDef::CloneCopyShim(def_id) => {
+            let substs = Substs::identity_for_item(tcx, def_id);
+            let self_ty = substs.type_at(0);
+            let mut builder = CloneShimBuilder::new(tcx, def_id, self_ty);
+            builder.copy_shim();
+            builder.into_mir()
```

Comment:
I..don't understand this. How can we have just one shim for all monomorphizations? Don't their sizes depend on the specific types in use?

Reply:
And what does this DefId refer to anyhow?

Reply:
Shims are not per-monomorphization, they're pretty generic (look at some of the other ones). `CloneShim` (and now the `CloneStructuralShim`) was *incorrectly* too monomorphic (but the array/tuple cases are harder to fix).

What's happening now is that there's *only one* MIR body for`return *self;` and one per closure *definition*, which you monomorphize just like you'd monomorphize a generic closure definition.

Reply:
"One shim for all monomorphizations" is the reason behind this PR. As for how it works, ask @eddyb